### DISCUSSION TITLE
**MacOS Ventura Truffle Installation Guide**

### DISCUSSION THREAD

**Discussion by apdan510** [ConsenSys-archive/truffle#5118] [Installation]
last tested: November 16, 2022

The steps below have been tested on:
**macOS 13.0.1 (Ventura)**
**macOS 12.3.1 (Monterey)**
**macOS 10.15.7 (Catalina)**

Requirements:
Node.js v14-v18 (NPM)

**Step 1 - Install Node.js**
To avoid permission errors when installing globally, Node.js (NPM) recommends installing Node.js with a [Node version manager](https://github.com/nvm-sh/nvm) (NVM) on Mac and Linux Operating systems

**Install a node version manager (NVM):**
Open a new terminal (you can find terminal in finder->utilities), enter this command to install NVM: 
`curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh | bash`

<img width="730" alt="image" src="https://user-images.githubusercontent.com/40121410/202536535-1199c80d-c0aa-4c00-a36a-47e5aa066a16.png">

<img width="945" alt="image" src="https://user-images.githubusercontent.com/40121410/202535544-5b89facc-cf37-41eb-8492-b6c4c1294923.png">

**Some users may encounter xcode-select developer tools error**

Some users may come across an error message (see below) when attempting to install NVM. If you are able to install NVM successfully without an error message please skip to step 2. 

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 15037  100 15037    0     0  65795      0 --:--:-- --:--:-- --:--:-- 67430
=> Downloading nvm from git to '/Users/matthew/.nvm'
=> xcode-select: note: no developer tools were found at '/Applications/Xcode.app', requesting install. Choose an option in the dialog to download the command line developer tools.
Failed to clone nvm repo. Please report this!
```

To resolve this issue run this command in the same terminal:
`xcode-select --install`

A message may pop up about installing command line tools. Click install and continue installation. 
<img width="579" alt="image" src="https://user-images.githubusercontent.com/40121410/202538785-7cd8111d-ce89-4488-bc1e-e5fa5b113818.png">

Try to install NVM again in the terminal with the same command:
`curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh | bash`

** Step 2 - Complete the NVM installation**

After the NVM installation is complete, verify installation by closing the terminal and opening a new terminal session. Some users may encounter this error message: 

`nvm: command not found `

Since macOS 10.15, the default shell is zsh and nvm will look for .zshrc to update, none is installed by default. Follow the steps below to resolve this issue.

Run this command in your terminal: 
`touch ~/.zshrc`

Run the NVM install script again:
`curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.2/install.sh | bash`

Verify installation by closing and reopening your terminal and test by entering this command: 
`nvm`

**Step 3 - install the latest Truffle compatible version of Node.js (NPM) in the terminal with:**
`nvm install 18`

<img width="944" alt="image" src="https://user-images.githubusercontent.com/40121410/202545175-0fa04810-08b9-452d-95df-6da5398064fd.png">

Verify that Npm is installed in your terminal with: `npm`

<img width="563" alt="image" src="https://user-images.githubusercontent.com/40121410/203664362-ed84d4e4-9c94-4baf-863e-4b1755024d5f.png">

_It is good practice to run the `npm doctor` command after you set up your Node.js environment on your computer. 
[Npm doctor](https://docs.npmjs.com/cli/v7/commands/npm-doctor) checks your Npm installation if the requirements to manage your JavaScript packages are configured and installed correctly. Since we are using a compatible version of NPM to install truffle (Node 18 , Npm 8) Npm doctor may alert you that your Npm installation to needs an update. This is because Npm doctor always looks at the latest version of Npm version available. You can ignore this warning and continue onto installing Truffle in the next step._ 

<img width="683" alt="image" src="https://user-images.githubusercontent.com/40121410/203664446-75fff68d-cb27-477d-b0a2-e70ac206f6c5.png">

**Step 4 - Install Truffle in a new terminal with:**
`npm install -g truffle`

<img width="614" alt="image" src="https://user-images.githubusercontent.com/40121410/202545631-942a0a75-6b06-4676-8260-7f6c36bbfdf2.png">

You may see some warnings or conflicts, this is typical and should work normally as long as there are no errors. 

**Test the Truffle install with:**
`truffle version`

<img width="567" alt="image" src="https://user-images.githubusercontent.com/40121410/202545706-439716c4-3c55-429f-a392-c2ac17dd91b0.png">

**Comment by hiteshgupta2507** [comment]
I still cannot get this to work.  After I have the .zshrc file created. I have the code below and saved the file.  
export NVM_DIR="$HOME/.nvm" . "/opt/homebrew/opt/nvm/nvm.sh"

When I run this command after nvm install --lts I get an error "bash: nvm: command not found" ... I have tried it with 
sudo nvm install --lts  and still the same error. What can I do to get truffle installed?

**Reply by apdan510** [reply]
Hey there,

Can you please try installing NVM again (does not need sudo) with:
`curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.1/install.sh | bash`

after it finishes, try testing it with just entering `NVM` and let me know what that returns.

Kind regards,
Alex

**Reply by hiteshgupta2507** [reply]
Hi Alex,

I was able to install truffle with lower node version. Initially I was on node >18.0.0 and I downgraded it to 14.0.0 and it worked.

**Comment by gaoshancha** [comment]
I can't get this to build on MacOS:

- macOS 12.4 Monterey
- Xcode 13.4.1
- node v16.15.1
- npm v8.11.0

I follow the instructions in the original posting, and I get the error messages, but don't seen any obvious error messages:

```
% npm install -g truffle
npm WARN config global '--global', '--local' are deprecated. Use '--location=global' instead.
npm WARN deprecated mkdirp-promise@5.0.1: This package is broken and no longer maintained. 'mkdirp' itself supports promises now, please switch to that.
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated uuid@2.0.1: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated multicodec@1.0.4: This module has been superseded by the multiformats module
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.2.1: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.2.1: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.3.2: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated multibase@0.6.1: This module has been superseded by the multiformats module
npm WARN deprecated multibase@0.7.0: This module has been superseded by the multiformats module
npm WARN deprecated multicodec@0.5.7: This module has been superseded by the multiformats module
npm WARN deprecated node-pre-gyp@0.11.0: Please upgrade to @mapbox/node-pre-gyp: the non-scoped node-pre-gyp package is deprecated and only the @mapbox scoped package will recieve updates in the future
npm WARN deprecated cids@0.7.5: This module has been superseded by the multiformats module

added 578 packages, and audited 579 packages in 3m

84 packages are looking for funding
  run 'npm fund' for details

4 high severity vulnerabilities

Some issues need review, and may require choosing
a different dependency.

Run 'npm audit' for details.
```

I tried to run the npm audit command:

```
% npm audit
found 0 vulnerabilities
```

I tried to just run truffle to see if it was installed. 

```
% truffle
zsh: permission denied: truffle
```

However, I didn't run the installation command with the sudo, so this seems odd. Using the find command, I found the file 'truffle' in '/usr/local/bin/truffle'.

```
% cd /usr/local/bin
% ls -al
...
lrwxr-xr-x   1 user  admin   48 Jun 21 20:16 truffle -> ../lib/node_modules/truffle/build/cli.bundled.js
...
```

However, ../lib/node_modules/truffle does not exist:

```
% ls -al ../lib/node_modules/
total 0
drwxr-xr-x    3 user  admin    96 Jun 21 20:17 .
drwxrwxr-x  169 user  admin  5408 Jun 21 20:22 ..
drwxr-xr-x   12 user  admin   384 Jun 21 20:11 npm

```

Is this indicative of the build failing with nvm? If so, where can I find the actual error so I can further debug?

**Reply by apdan510** [reply]
@gaoshancha NVM should be handling these permission issues. Where does your global /node_modules folder live? Is it in usr/local/bin?

Kind regards,
Alex | Truffle Support

**Reply by hiteshgupta2507** [reply]
@gaoshancha if you have nvm installed can you try `nvm use 14` then do `npm install -g truffle`

It worked for me.

**Reply by gaoshancha** [reply]
> @gaoshancha NVM should be handling these permission issues. Where does your global /node_modules folder live? Is it in usr/local/bin?
 
It seems to be in /usr/local/lib/node_modules/npm/node_modules:

```
% pwd
/usr/local/lib/node_modules/npm/node_modules
 % ls
@colors				debug				ip-regex			mute-stream			rimraf
@gar				debuglog			is-cidr				negotiator			safe-buffer
@isaacs				defaults			is-core-module			node-gyp			safer-buffer
@npmcli				delegates			is-fullwidth-code-point		nopt				semver
@tootallnate			depd				is-lambda			normalize-package-data		set-blocking
abbrev				dezalgo				isexe				npm-audit-report		signal-exit
agent-base			diff				json-parse-even-better-errors	npm-bundled			smart-buffer
agentkeepalive			emoji-regex			json-stringify-nice		npm-install-checks		socks
aggregate-error			encoding			jsonparse			npm-normalize-package-bin	socks-proxy-agent
ansi-regex			env-paths			just-diff			npm-package-arg			spdx-correct
ansi-styles			err-code			just-diff-apply			npm-packlist			spdx-exceptions
aproba				fastest-levenshtein		libnpmaccess			npm-pick-manifest		spdx-expression-parse
archy				fs-minipass			libnpmdiff			npm-profile			spdx-license-ids
are-we-there-yet		fs.realpath			libnpmexec			npm-registry-fetch		ssri
asap				function-bind			libnpmfund			npm-user-validate		string-width
balanced-match			gauge				libnpmhook			npmlog				string_decoder
bin-links			glob				libnpmorg			once				strip-ansi
binary-extensions		graceful-fs			libnpmpack			opener				supports-color
brace-expansion			has				libnpmpublish			p-map				tar
builtins			has-flag			libnpmsearch			pacote				text-table
cacache				has-unicode			libnpmteam			parse-conflict-json		tiny-relative-date
chalk				hosted-git-info			libnpmversion			path-is-absolute		treeverse
chownr				http-cache-semantics		lru-cache			proc-log			unique-filename
cidr-regex			http-proxy-agent		make-fetch-happen		promise-all-reject-late		unique-slug
clean-stack			https-proxy-agent		minimatch			promise-call-limit		util-deprecate
cli-columns			humanize-ms			minipass			promise-inflight		validate-npm-package-license
cli-table3			iconv-lite			minipass-collect		promise-retry			validate-npm-package-name
clone				ignore-walk			minipass-fetch			promzard			walk-up-path
cmd-shim			imurmurhash			minipass-flush			qrcode-terminal			wcwidth
color-convert			indent-string			minipass-json-stream		read				which
color-name			infer-owner			minipass-pipeline		read-cmd-shim			wide-align
color-support			inflight			minipass-sized			read-package-json		wrappy
columnify			inherits			minizlib			read-package-json-fast		write-file-atomic
common-ancestor-path		ini				mkdirp				readable-stream			yallist
concat-map			init-package-json		mkdirp-infer-owner		readdir-scoped-modules
console-control-strings		ip				ms				retry
```

The parent folder /usr/local/lib/node_modules/npm has the npm bin, docs, etc:

 ```
% ls
LICENSE		README.md	bin		docs		index.js	lib		man		node_modules	npmrc		package.json
```

**Reply by apdan510** [reply]
can we try downgrading your node version?

**Downgrade your node version using NVM to Node 14 and npm 6:**
**In a terminal:**
`nvm install 14`
**then**
`nvm use 14.19.3`

**then try installing truffle one more time**
`npm install -g truffle`

**Reply by gaoshancha** [reply]
Downgrading seems to do the trick as you both mentioned. It looks like node@16 and node@18 are both incompatible with Truffle.

```
% nvm use 14
Now using node v14.19.3 (npm v6.14.17)

% truffle
Truffle v5.5.18 - a development framework for Ethereum

Usage: truffle <command> [options]

Commands:
  truffle build      Execute build pipeline (if configuration present)
  truffle compile    Compile contract source files
  truffle config     Set user-level configuration options
  truffle console    Run a console with contract abstractions and commands
                     available
  truffle create     Helper to create new contracts, migrations and tests
  truffle dashboard  Start Truffle Dashboard to sign development transactions
                     using browser wallet
  truffle db         Database interface commands
  truffle debug      Interactively debug any transaction on the blockchain
  truffle deploy     (alias for migrate)
  truffle develop    Open a console with a local development blockchain
  truffle exec       Execute a JS module within this Truffle environment
  truffle help       List all commands or provide information about a specific
                     command
  truffle init       Initialize new and empty Ethereum project
  truffle install    Install a package from the Ethereum Package Registry
  truffle migrate    Run migrations to deploy contracts
  truffle networks   Show addresses for deployed contracts on each network
  truffle obtain     Fetch and cache a specified compiler
  truffle opcode     Print the compiled opcodes for a given contract
  truffle preserve   Save data to decentralized storage platforms like IPFS and
                     Filecoin
  truffle publish    Publish a package to the Ethereum Package Registry
  truffle run        Run a third-party command
  truffle test       Run JavaScript and Solidity tests
  truffle unbox      Download a Truffle Box, a pre-built Truffle project
  truffle version    Show version number and exit
  truffle watch      Watch filesystem for changes and rebuild the project
                     automatically

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]

See more at http://trufflesuite.com/docs
```

**Reply by hiteshgupta2507** [reply]
@gaoshancha Amazing. Glad it worked.

**Comment by hiteshgupta2507** [comment]
Hi,
Seems like you have a different python version installed on your machine and you are trying to run the code with higher python version. 
1)You can upgrade your python version2) If the step 1 doesn't solve the issue check your node version if it's over 14.0.0 then downgrade it to 14 using nvm use 14
In a terminal:
nvm install 14
then
nvm use 14.19.3

then try installing truffle one more time

npm install -g truffle


    On Saturday, 9 July, 2022 at 11:40:45 pm GMT-4, holywatr ***@***.***> wrote:  
 
 


Hello ! I've been trying for hours to get Truffle onto my Mac OS Monterey, and have had zero luck. I'm brand new in the dev world so that might have something to do with it - but I'm hoping someone here can help me :/

I've downloaded Node / NPM / NVM ... done everything I've read to do - and yet I still receive erros.

So I type in:
npm i -g truffle

and it all goes smoothly up until:

npm WARN deprecated ***@***.***: This module has been superseded by the multiformats module

at which point, I start receiving erros that look like this:

npm ERR! code 1 npm ERR! path /Users/patrickmiddelthon/.nvm/versions/node/v18.5.0/lib/node_modules/truffle/node_modules/ganache/node_modules/leveldown npm ERR! command failed npm ERR! command sh -c node-gyp rebuild npm ERR! CXX(target) Release/obj.target/leveldb/deps/leveldb/leveldb-1.20/db/builder.o npm ERR! gyp info it worked if it ends with ok npm ERR! gyp info using ***@***.*** npm ERR! gyp info using ***@***.*** | darwin | x64 npm ERR! gyp info find Python using Python version 3.8.2 found at "/Library/Developer/CommandLineTools/usr/bin/python3" npm ERR! gyp http GET https://nodejs.org/download/release/v18.5.0/node-v18.5.0-headers.tar.gz npm ERR! gyp http 200 https://nodejs.org/download/release/v18.5.0/node-v18.5.0-headers.tar.gz npm ERR! gyp http GET https://nodejs.org/download/release/v18.5.0/SHASUMS256.txt npm ERR! gyp http 200 https://nodejs.org/download/release/v18.5.0/SHASUMS256.txt npm ERR! gyp info spawn /Library/Developer/CommandLineTools/usr/bin/python3 

And the list goes on, and on, and on :)

So, I have no idea what I'm looking at, or what to do - and I'm mentally exhausted from figuring all this out by myself haha YouTube sure helps a lot though. Can someone help me troubleshoot this please? All I want to do is install Truffle haha I want to create Dapps and bots and fun stuff to explore Web3 with !!
